### PR TITLE
use BrowserRouter to add basename and adjust Cypress tests to be aligned

### PR DESCRIPTION
**Pull request by ElevnLi**
### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by ElevnLi** [comment]
This is really neat to resolve most of what we need, but there are some tests it can't go through
Cypress results for this tweak:
[cypress results.txt](https://github.com/Weed-AI/cvat/files/7851129/cypress.results.txt)

**Comment by jnothman** [comment]
Not sure if your comment belongs here.

File: cvat-core/src/api.js [hunk 1]
```diff
@@ -711,6 +712,12 @@ function build() {
             set proxy(value) {
                 config.proxy = value;
             },
+            get prefix() {
```
Comment (jnothman) [inline cvat-core/src/api.js]: should we call this `basename`?

File: cvat-ui/src/cvat-core-wrapper.ts [hunk 2]
```diff
@@ -5,8 +5,10 @@
 import _cvat from 'cvat-core/src/api';
 
 const cvat: any = _cvat;
+const prefix: string = '/cvat-annotation';
```
Comment (jnothman) [inline cvat-ui/src/cvat-core-wrapper.ts]: I'm confused about why this config is set here rather than in config.js...

**Review by jnothman — COMMENTED** [review]
Was cypress all passing on #2?

**Comment by ElevnLi** [comment]
I've just done a full test without the prefix.
From [cypress.results with prefix.txt](https://github.com/Weed-AI/cvat/files/7851243/cypress.results.with.prefix.txt), I got 36 of 138 failed tests (26%), and 89 of 659 failed steps.
From [cypress.results without prefix.txt](https://github.com/Weed-AI/cvat/files/7851247/cypress.results.without.prefix.txt), I got 10 of 138 failed tests (7%), and 18 of 659 failed steps.
So some manual tweaks need to be done to make it as functional as the original version, especially something like registration doesn't work with the prefix.

File: cvat-ui/src/cvat-core-wrapper.ts (hunk 2 above)
Comment (ElevnLi) [reply cvat-ui/src/cvat-core-wrapper.ts]: The cvat-ui doesn't have a config.js file for now. It only exists in cvat-core dir. ofc we can create one.

**Comment by ElevnLi** [comment]
The reason I added baseUrl in cypress.json and manually applied basename as env variable to some tests is that [baseUrl](https://docs.cypress.io/guides/references/configuration#cypress-json) only supports cy.visit and cy.request, so things like a[href] and cy.intercept need manual adjustment.

**Comment by ElevnLi** [comment]
[cypress.results with prefix after changing cypress config.txt](https://github.com/Weed-AI/cvat/files/7852350/cypress.results.with.prefix.after.changing.cypress.config.txt)
Now the passing rate is about the same as without basename.

File: cvat-core/src/api.js [hunk 3]
```diff
@@ -697,6 +697,7 @@ function build() {
              * @property {string} proxy Axios proxy settings.
              * For more details please read <a href="https://github.com/axios/axios"> here </a>
              * @memberof module:API.cvat.config
+             * @property {string} basename
```
Comment (jnothman) [inline cvat-core/src/api.js]: Document its function?

Perhaps it should be `basePath`?

**Review by jnothman — COMMENTED** [review]
Nice work :)


I wonder if you can directly use the existing `baseUrl` setting in `cy.intercept` and `a[href]` even if it's not automatic.

File: cvat-core/src/api.js (hunk 3 above)
Comment (ElevnLi) [reply cvat-core/src/api.js]: baseUrl is not accessible as env variable from where I can see, so probably have to keep them both.
Do we want to rename it to basePath? It's a bit of work as there are basename used some other dirs.

**Review by jnothman — APPROVED** [review]
Let's go with this :)

**Pull request merged**